client/consensus/babe/src/tests.rs

```diff
@@ -894,3 +897,76 @@ fn babe_transcript_generation_match() {
 	};
 	debug_assert!(test(orig_transcript) == test(transcript_from_data(new_transcript)));
 }
+
+#[test]
+fn obsolete_blocks_aux_data_cleanup() {
+	let mut net = BabeTestNet::new(1);
+
+	let peer = net.peer(0);
+	let data = peer.data.as_ref().expect("babe link set up during initialization");
+	let client = peer.client().as_client();
+
+	let mut proposer_factory = DummyFactory {
+		client: client.clone(),
+		config: data.link.config.clone(),
+		epoch_changes: data.link.epoch_changes.clone(),
+		mutator: Arc::new(|_, _| ()),
+	};
+
+	let mut block_import = data.block_import.lock().take().expect("import set up during init");
+
+	let mut propose_and_import_blocks_wrap = |parent_id, n| {
+		propose_and_import_blocks(&client, &mut proposer_factory, &mut block_import, parent_id, n)
+	};
+
+	let aux_data_check = |hashes: &[Hash], expected: bool| {
+		hashes.iter().all(|hash| {
+			aux_schema::load_block_weight(&*peer.client().as_backend(), hash)
+				.unwrap()
+				.is_some() == expected
+		})
+	};
+
+	// Create the following test scenario:
+	//
+	//               /-----B3 --- B4           ( < fork2 )
+	// G --- A1 --- A2 --- A3 --- A4           ( < fork1 )
+	//                      \-----C4 --- C5    ( < fork3 )
+
+	let fork1_hashes = propose_and_import_blocks_wrap(BlockId::Number(0), 4);
+	let fork2_hashes = propose_and_import_blocks_wrap(BlockId::Number(2), 2);
+	let fork3_hashes = propose_and_import_blocks_wrap(BlockId::Number(3), 2);
+
+	// Check that aux data is present for all but the genesis block.
+	assert!(aux_data_check(&[client.chain_info().genesis_hash], false));
+	assert!(aux_data_check(&fork1_hashes, true));
+	assert!(aux_data_check(&fork2_hashes, true));
+	assert!(aux_data_check(&fork3_hashes, true));
+
+	// Trigger cleanup. Actions order is important:
+	// 1. Get the finality notification stream to create a sink entry within the client.
+	// 2. Finalize a block (A2). This will send finality notifications to subscribers.
+	// 3. Corcively close the finality notification sinks to allow `aux_storage_cleanup` future to
+	//    exit as soon as the outstanding finality messages are consumed.
+
+	let finality_notifications = client.finality_notification_stream();
+
+	client.finalize_block(BlockId::Number(3), None, true).unwrap();
+
+	block_on(async {
+		for mut sink in client.finality_notification_sinks().lock().iter() {
+			let _ = sink.close().await;
```

Comment:
```suggestion
			// Close the sinks so that `aux_storage_cleanup` will end after is has processed all notifications.
			let _ = sink.close().await;
```